Unsatisfied dependency for type io.vertx.db2client.DB2Pool and qualifiers with Quarkus master, works with 1.8.0.Final

I see Unsatisfied dependency for type io.vertx.db2client.DB2Pool and qualifiers with Quarkus master (6588419c63) when running daily GH CI check on https://github.com/quarkus-qe/beefy-scenarios. Works well with 1.8.0.Final.

```java
java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.arc.deployment.ArcProcessor#validate threw an exception: javax.enterprise.inject.spi.DeploymentException: javax.enterprise.inject.UnsatisfiedResolutionException: Unsatisfied dependency for type io.vertx.db2client.DB2Pool and qualifiers [@Default]
	- java member: io.quarkus.reactive.db2.client.runtime.DB2PoolProducer#db2Pool
	- declared on CLASS bean [types=[java.lang.Object, io.quarkus.reactive.db2.client.runtime.DB2PoolProducer], qualifiers=[@Default, @Any], target=io.quarkus.reactive.db2.client.runtime.DB2PoolProducer]
	at io.quarkus.arc.processor.BeanDeployment.processErrors(BeanDeployment.java:1018)
	at io.quarkus.arc.processor.BeanDeployment.init(BeanDeployment.java:238)
	at io.quarkus.arc.processor.BeanProcessor.initialize(BeanProcessor.java:122)
	at io.quarkus.arc.deployment.ArcProcessor.validate(ArcProcessor.java:393)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:936)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
	at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
	at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:2046)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1578)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1452)
	at java.base/java.lang.Thread.run(Thread.java:834)
	at org.jboss.threads.JBossThread.run(JBossThread.java:479)
Caused by: javax.enterprise.inject.UnsatisfiedResolutionException: Unsatisfied dependency for type io.vertx.db2client.DB2Pool and qualifiers [@Default]
	- java member: io.quarkus.reactive.db2.client.runtime.DB2PoolProducer#db2Pool
	- declared on CLASS bean [types=[java.lang.Object, io.quarkus.reactive.db2.client.runtime.DB2PoolProducer], qualifiers=[@Default, @Any], target=io.quarkus.reactive.db2.client.runtime.DB2PoolProducer]
	at io.quarkus.arc.processor.Beans.resolveInjectionPoint(Beans.java:504)
	at io.quarkus.arc.processor.BeanInfo.init(BeanInfo.java:363)
	at io.quarkus.arc.processor.BeanDeployment.init(BeanDeployment.java:230)
	... 14 more

```

Reproducer (expecting Quarkus master was built beforehand):
```bash
git clone https://github.com/quarkus-qe/beefy-scenarios.git
cd beefy-scenarios
mvn clean verify -pl 002-quarkus-all-extensions  
```

Running with Quarkus 1.8.0.Final works fine (`mvn clean verify -pl 002-quarkus-all-extensions/ -Dquarkus.platform.version=1.8.0.Final -Dquarkus-plugin.version=1.8.0.Final`).

--- a/extensions/reactive-db2-client/deployment/src/main/java/io/quarkus/reactive/db2/client/deployment/ReactiveDB2ClientProcessor.java
+++ b/extensions/reactive-db2-client/deployment/src/main/java/io/quarkus/reactive/db2/client/deployment/ReactiveDB2ClientProcessor.java
@@ -4,6 +4,7 @@ import javax.inject.Singleton;
 
 import io.quarkus.arc.deployment.AdditionalBeanBuildItem;
 import io.quarkus.arc.deployment.SyntheticBeanBuildItem;
+import io.quarkus.arc.processor.DotNames;
 import io.quarkus.datasource.common.runtime.DatabaseKind;
 import io.quarkus.datasource.runtime.DataSourcesBuildTimeConfig;
 import io.quarkus.datasource.runtime.DataSourcesRuntimeConfig;
@@ -30,9 +31,24 @@ import io.vertx.sqlclient.Pool;
 
 class ReactiveDB2ClientProcessor {
 
+    /**
+     * The producer needs to be produced in a separate method to avoid a circular dependency (the Vert.x instance creation
+     * consumes the AdditionalBeanBuildItems).
+     */
     @BuildStep
-    AdditionalBeanBuildItem registerBean() {
-        return AdditionalBeanBuildItem.unremovableOf(DB2PoolProducer.class);
+    void poolProducer(
+            BuildProducer<AdditionalBeanBuildItem> additionalBeans,
+            DataSourcesBuildTimeConfig dataSourcesBuildTimeConfig,
+            DataSourceReactiveBuildTimeConfig dataSourceReactiveBuildTimeConfig) {
+        if (!createPool(dataSourcesBuildTimeConfig, dataSourceReactiveBuildTimeConfig)) {
+            return;
+        }
+
+        additionalBeans.produce(new AdditionalBeanBuildItem.Builder()
+                .addBeanClass(DB2PoolProducer.class)
+                .setUnremovable()
+                .setDefaultScope(DotNames.APPLICATION_SCOPED)
+                .build());
     }
 
     @BuildStep
@@ -53,12 +69,7 @@ class ReactiveDB2ClientProcessor {
         // Make sure the DB2PoolProducer is initialized before the StartupEvent is fired
         ServiceStartBuildItem serviceStart = new ServiceStartBuildItem("reactive-db2-client");
 
-        if (!dataSourcesBuildTimeConfig.defaultDataSource.dbKind.isPresent()) {
-            return serviceStart;
-        }
-
-        if (!DatabaseKind.isDB2(dataSourcesBuildTimeConfig.defaultDataSource.dbKind.get())
-                || !dataSourceReactiveBuildTimeConfig.enabled) {
+        if (!createPool(dataSourcesBuildTimeConfig, dataSourceReactiveBuildTimeConfig)) {
             return serviceStart;
         }
 
@@ -82,8 +93,30 @@ class ReactiveDB2ClientProcessor {
     }
 
     @BuildStep
-    HealthBuildItem addHealthCheck(DataSourcesBuildTimeConfig dataSourcesBuildTimeConfig) {
-        return new HealthBuildItem("io.quarkus.reactive.db2.client.runtime.health.ReactiveDB2DataSourceHealthCheck",
-                dataSourcesBuildTimeConfig.healthEnabled);
+    void addHealthCheck(
+            BuildProducer<HealthBuildItem> healthChecks,
+            DataSourcesBuildTimeConfig dataSourcesBuildTimeConfig,
+            DataSourceReactiveBuildTimeConfig dataSourceReactiveBuildTimeConfig) {
+        if (!createPool(dataSourcesBuildTimeConfig, dataSourceReactiveBuildTimeConfig)) {
+            return;
+        }
+
+        healthChecks
+                .produce(new HealthBuildItem("io.quarkus.reactive.db2.client.runtime.health.ReactiveDB2DataSourceHealthCheck",
+                        dataSourcesBuildTimeConfig.healthEnabled));
+    }
+
+    private static boolean createPool(DataSourcesBuildTimeConfig dataSourcesBuildTimeConfig,
+            DataSourceReactiveBuildTimeConfig dataSourceReactiveBuildTimeConfig) {
+        if (!dataSourcesBuildTimeConfig.defaultDataSource.dbKind.isPresent()) {
+            return false;
+        }
+
+        if (!DatabaseKind.isDB2(dataSourcesBuildTimeConfig.defaultDataSource.dbKind.get())
+                || !dataSourceReactiveBuildTimeConfig.enabled) {
+            return false;
+        }
+
+        return true;
     }
 }

--- a/extensions/reactive-db2-client/runtime/src/main/java/io/quarkus/reactive/db2/client/runtime/DB2PoolProducer.java
+++ b/extensions/reactive-db2-client/runtime/src/main/java/io/quarkus/reactive/db2/client/runtime/DB2PoolProducer.java
@@ -1,13 +1,11 @@
 package io.quarkus.reactive.db2.client.runtime;
 
-import javax.enterprise.context.ApplicationScoped;
 import javax.enterprise.inject.Produces;
 import javax.inject.Inject;
 import javax.inject.Singleton;
 
 import io.vertx.db2client.DB2Pool;
 
-@ApplicationScoped
 public class DB2PoolProducer {
 
     @Inject

--- a/extensions/reactive-mysql-client/deployment/src/main/java/io/quarkus/reactive/mysql/client/deployment/ReactiveMySQLClientProcessor.java
+++ b/extensions/reactive-mysql-client/deployment/src/main/java/io/quarkus/reactive/mysql/client/deployment/ReactiveMySQLClientProcessor.java
@@ -4,6 +4,7 @@ import javax.inject.Singleton;
 
 import io.quarkus.arc.deployment.AdditionalBeanBuildItem;
 import io.quarkus.arc.deployment.SyntheticBeanBuildItem;
+import io.quarkus.arc.processor.DotNames;
 import io.quarkus.datasource.common.runtime.DatabaseKind;
 import io.quarkus.datasource.runtime.DataSourcesBuildTimeConfig;
 import io.quarkus.datasource.runtime.DataSourcesRuntimeConfig;
@@ -28,12 +29,26 @@ import io.quarkus.vertx.deployment.VertxBuildItem;
 import io.vertx.mysqlclient.MySQLPool;
 import io.vertx.sqlclient.Pool;
 
-@SuppressWarnings("deprecation")
 class ReactiveMySQLClientProcessor {
 
+    /**
+     * The producer needs to be produced in a separate method to avoid a circular dependency (the Vert.x instance creation
+     * consumes the AdditionalBeanBuildItems).
+     */
     @BuildStep
-    AdditionalBeanBuildItem registerBean() {
-        return AdditionalBeanBuildItem.unremovableOf(MySQLPoolProducer.class);
+    void poolProducer(
+            BuildProducer<AdditionalBeanBuildItem> additionalBeans,
+            DataSourcesBuildTimeConfig dataSourcesBuildTimeConfig,
+            DataSourceReactiveBuildTimeConfig dataSourceReactiveBuildTimeConfig) {
+        if (!createPool(dataSourcesBuildTimeConfig, dataSourceReactiveBuildTimeConfig)) {
+            return;
+        }
+
+        additionalBeans.produce(new AdditionalBeanBuildItem.Builder()
+                .addBeanClass(MySQLPoolProducer.class)
+                .setUnremovable()
+                .setDefaultScope(DotNames.APPLICATION_SCOPED)
+                .build());
     }
 
     @BuildStep
@@ -54,13 +69,7 @@ class ReactiveMySQLClientProcessor {
         // Make sure the MySQLPoolProducer is initialized before the StartupEvent is fired
         ServiceStartBuildItem serviceStart = new ServiceStartBuildItem("reactive-mysql-client");
 
-        if (!dataSourcesBuildTimeConfig.defaultDataSource.dbKind.isPresent()) {
-            return serviceStart;
-        }
-
-        if ((!DatabaseKind.isMySQL(dataSourcesBuildTimeConfig.defaultDataSource.dbKind.get())
-                && !DatabaseKind.isMariaDB(dataSourcesBuildTimeConfig.defaultDataSource.dbKind.get()))
-                || !dataSourceReactiveBuildTimeConfig.enabled) {
+        if (!createPool(dataSourcesBuildTimeConfig, dataSourceReactiveBuildTimeConfig)) {
             return serviceStart;
         }
 
@@ -84,8 +93,31 @@ class ReactiveMySQLClientProcessor {
     }
 
     @BuildStep
-    HealthBuildItem addHealthCheck(DataSourcesBuildTimeConfig dataSourcesBuildTimeConfig) {
-        return new HealthBuildItem("io.quarkus.reactive.mysql.client.runtime.health.ReactiveMySQLDataSourceHealthCheck",
-                dataSourcesBuildTimeConfig.healthEnabled);
+    void addHealthCheck(
+            BuildProducer<HealthBuildItem> healthChecks,
+            DataSourcesBuildTimeConfig dataSourcesBuildTimeConfig,
+            DataSourceReactiveBuildTimeConfig dataSourceReactiveBuildTimeConfig) {
+        if (!createPool(dataSourcesBuildTimeConfig, dataSourceReactiveBuildTimeConfig)) {
+            return;
+        }
+
+        healthChecks.produce(
+                new HealthBuildItem("io.quarkus.reactive.mysql.client.runtime.health.ReactiveMySQLDataSourceHealthCheck",
+                        dataSourcesBuildTimeConfig.healthEnabled));
+    }
+
+    private static boolean createPool(DataSourcesBuildTimeConfig dataSourcesBuildTimeConfig,
+            DataSourceReactiveBuildTimeConfig dataSourceReactiveBuildTimeConfig) {
+        if (!dataSourcesBuildTimeConfig.defaultDataSource.dbKind.isPresent()) {
+            return false;
+        }
+
+        if ((!DatabaseKind.isMySQL(dataSourcesBuildTimeConfig.defaultDataSource.dbKind.get())
+                && !DatabaseKind.isMariaDB(dataSourcesBuildTimeConfig.defaultDataSource.dbKind.get()))
+                || !dataSourceReactiveBuildTimeConfig.enabled) {
+            return false;
+        }
+
+        return true;
     }
 }

--- a/extensions/reactive-mysql-client/runtime/src/main/java/io/quarkus/reactive/mysql/client/runtime/MySQLPoolProducer.java
+++ b/extensions/reactive-mysql-client/runtime/src/main/java/io/quarkus/reactive/mysql/client/runtime/MySQLPoolProducer.java
@@ -1,6 +1,5 @@
 package io.quarkus.reactive.mysql.client.runtime;
 
-import javax.enterprise.context.ApplicationScoped;
 import javax.enterprise.inject.Produces;
 import javax.inject.Inject;
 import javax.inject.Singleton;
@@ -9,7 +8,6 @@ import org.jboss.logging.Logger;
 
 import io.vertx.mysqlclient.MySQLPool;
 
-@ApplicationScoped
 public class MySQLPoolProducer {
 
     private static final Logger LOGGER = Logger.getLogger(MySQLPoolProducer.class);

--- a/extensions/reactive-pg-client/deployment/src/main/java/io/quarkus/reactive/pg/client/deployment/ReactivePgClientProcessor.java
+++ b/extensions/reactive-pg-client/deployment/src/main/java/io/quarkus/reactive/pg/client/deployment/ReactivePgClientProcessor.java
@@ -4,6 +4,7 @@ import javax.inject.Singleton;
 
 import io.quarkus.arc.deployment.AdditionalBeanBuildItem;
 import io.quarkus.arc.deployment.SyntheticBeanBuildItem;
+import io.quarkus.arc.processor.DotNames;
 import io.quarkus.datasource.common.runtime.DatabaseKind;
 import io.quarkus.datasource.runtime.DataSourcesBuildTimeConfig;
 import io.quarkus.datasource.runtime.DataSourcesRuntimeConfig;
@@ -29,12 +30,26 @@ import io.quarkus.vertx.deployment.VertxBuildItem;
 import io.vertx.pgclient.PgPool;
 import io.vertx.sqlclient.Pool;
 
-@SuppressWarnings("deprecation")
 class ReactivePgClientProcessor {
 
+    /**
+     * The producer needs to be produced in a separate method to avoid a circular dependency (the Vert.x instance creation
+     * consumes the AdditionalBeanBuildItems).
+     */
     @BuildStep
-    AdditionalBeanBuildItem registerBean() {
-        return AdditionalBeanBuildItem.unremovableOf(PgPoolProducer.class);
+    void poolProducer(
+            BuildProducer<AdditionalBeanBuildItem> additionalBeans,
+            DataSourcesBuildTimeConfig dataSourcesBuildTimeConfig,
+            DataSourceReactiveBuildTimeConfig dataSourceReactiveBuildTimeConfig) {
+        if (!createPool(dataSourcesBuildTimeConfig, dataSourceReactiveBuildTimeConfig)) {
+            return;
+        }
+
+        additionalBeans.produce(new AdditionalBeanBuildItem.Builder()
+                .addBeanClass(PgPoolProducer.class)
+                .setUnremovable()
+                .setDefaultScope(DotNames.APPLICATION_SCOPED)
+                .build());
     }
 
     @BuildStep
@@ -62,12 +77,7 @@ class ReactivePgClientProcessor {
         // Make sure the PgPoolProducer is initialized before the StartupEvent is fired
         ServiceStartBuildItem serviceStart = new ServiceStartBuildItem("reactive-pg-client");
 
-        if (!dataSourcesBuildTimeConfig.defaultDataSource.dbKind.isPresent()) {
-            return serviceStart;
-        }
-
-        if (!DatabaseKind.isPostgreSQL(dataSourcesBuildTimeConfig.defaultDataSource.dbKind.get())
-                || !dataSourceReactiveBuildTimeConfig.enabled) {
+        if (!createPool(dataSourcesBuildTimeConfig, dataSourceReactiveBuildTimeConfig)) {
             return serviceStart;
         }
 
@@ -91,8 +101,29 @@ class ReactivePgClientProcessor {
     }
 
     @BuildStep
-    HealthBuildItem addHealthCheck(DataSourcesBuildTimeConfig dataSourcesBuildTimeConfig) {
-        return new HealthBuildItem("io.quarkus.reactive.pg.client.runtime.health.ReactivePgDataSourceHealthCheck",
-                dataSourcesBuildTimeConfig.healthEnabled);
+    void addHealthCheck(
+            BuildProducer<HealthBuildItem> healthChecks,
+            DataSourcesBuildTimeConfig dataSourcesBuildTimeConfig,
+            DataSourceReactiveBuildTimeConfig dataSourceReactiveBuildTimeConfig) {
+        if (!createPool(dataSourcesBuildTimeConfig, dataSourceReactiveBuildTimeConfig)) {
+            return;
+        }
+
+        healthChecks.produce(new HealthBuildItem("io.quarkus.reactive.pg.client.runtime.health.ReactivePgDataSourceHealthCheck",
+                dataSourcesBuildTimeConfig.healthEnabled));
+    }
+
+    private static boolean createPool(DataSourcesBuildTimeConfig dataSourcesBuildTimeConfig,
+            DataSourceReactiveBuildTimeConfig dataSourceReactiveBuildTimeConfig) {
+        if (!dataSourcesBuildTimeConfig.defaultDataSource.dbKind.isPresent()) {
+            return false;
+        }
+
+        if (!DatabaseKind.isPostgreSQL(dataSourcesBuildTimeConfig.defaultDataSource.dbKind.get())
+                || !dataSourceReactiveBuildTimeConfig.enabled) {
+            return false;
+        }
+
+        return true;
     }
 }

--- a/extensions/reactive-pg-client/runtime/src/main/java/io/quarkus/reactive/pg/client/runtime/PgPoolProducer.java
+++ b/extensions/reactive-pg-client/runtime/src/main/java/io/quarkus/reactive/pg/client/runtime/PgPoolProducer.java
@@ -1,6 +1,5 @@
 package io.quarkus.reactive.pg.client.runtime;
 
-import javax.enterprise.context.ApplicationScoped;
 import javax.enterprise.inject.Produces;
 import javax.inject.Inject;
 import javax.inject.Singleton;
@@ -9,7 +8,6 @@ import org.jboss.logging.Logger;
 
 import io.vertx.pgclient.PgPool;
 
-@ApplicationScoped
 public class PgPoolProducer {
 
     private static final Logger LOGGER = Logger.getLogger(PgPoolProducer.class);

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/DotNames.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/DotNames.java
@@ -12,6 +12,7 @@ import java.util.OptionalLong;
 import javax.annotation.PostConstruct;
 import javax.annotation.PreDestroy;
 import javax.annotation.Priority;
+import javax.enterprise.context.ApplicationScoped;
 import javax.enterprise.context.Initialized;
 import javax.enterprise.context.control.ActivateRequestContext;
 import javax.enterprise.event.Event;
@@ -82,6 +83,7 @@ public final class DotNames {
     public static final DotName ALTERNATIVE_PRIORITY = create(AlternativePriority.class);
     public static final DotName DEFAULT_BEAN = create(DefaultBean.class);
     public static final DotName SINGLETON = create(Singleton.class);
+    public static final DotName APPLICATION_SCOPED = create(ApplicationScoped.class);
     public static final DotName STEREOTYPE = create(Stereotype.class);
     public static final DotName TYPED = create(Typed.class);
     public static final DotName VETOED = create(Vetoed.class);
@@ -117,7 +119,7 @@ public final class DotNames {
      * Note that this method does not attempt to detect a nested class because the computing cache is shared with the
      * {@link #create(String)} variant and so the results would be inconsistent. Therefore, this method should only be used for
      * top-level classes.
-     * 
+     *
      * @param clazz
      * @return the computed dot name
      */
@@ -129,7 +131,7 @@ public final class DotNames {
      * Note that the dollar sign is a valid character for class names so we cannot detect a nested class here. Therefore, this
      * method returns a dot name for which {@link DotName#local()} returns {@code Foo$Bar} for the parameter
      * "com.foo.Foo$Bar".
-     * 
+     *
      * @param name
      * @return the computed dot name
      */
@@ -145,7 +147,7 @@ public final class DotNames {
     }
 
     /**
-     * 
+     *
      * @param clazz
      * @return the simple name for the given top-level or nested class
      */
@@ -175,7 +177,7 @@ public final class DotNames {
      * method returns "Foo$Bar" for the parameter "com.foo.Foo$Bar". Use {@link #simpleName(ClassInfo)} when you need to
      * distinguish
      * the nested classes.
-     * 
+     *
      * @param name
      * @return the simple name
      */

